File: cpp_code/real_numbers/real_numbers_examples.cc
// cpp_code/real_numbers/real_numbers_examples.cc
#include "real_numbers_utils.h" // Includes real_numbers_types.h as well
#include <iostream>
#include <vector>
#include <stdexcept> // For catching exceptions

int main() {
    std::cout << "--- Real Numbers Examples ---" << std::endl;

    // 1. Euclid's Division Lemma
    std::cout << "\n1. Euclid's Division Lemma:" << std::endl;
    try {
        RealNumbers::EuclidLemmaResult lemma1 = RealNumbers::euclidsDivisionLemma(455, 42);
        std::cout << "   455 / 42: " << lemma1.toString() << std::endl;
        RealNumbers::EuclidLemmaResult lemma2 = RealNumbers::euclidsDivisionLemma(-7, 3);
        std::cout << "   -7 / 3: " << lemma2.toString() << std::endl;
        RealNumbers::EuclidLemmaResult lemma3 = RealNumbers::euclidsDivisionLemma(7, -3);
        std::cout << "   7 / -3: " << lemma3.toString() << std::endl;
        RealNumbers::EuclidLemmaResult lemma4 = RealNumbers::euclidsDivisionLemma(-7, -3);
        std::cout << "   -7 / -3: " << lemma4.toString() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error in Euclid's Lemma: " << e.what() << std::endl;
    }

    // 2. Euclid's Algorithm for HCF
    std::cout << "\n2. Euclid's Algorithm for HCF:" << std::endl;
    try {
        RealNumbers::HCFResult hcf1 = RealNumbers::euclidsAlgorithmHCF(455, 42);
        std::cout << "   HCF(455, 42): " << hcf1.toString() << std::endl;
        RealNumbers::HCFResult hcf2 = RealNumbers::euclidsAlgorithmHCF(96, 404);
        std::cout << "   HCF(96, 404): " << hcf2.toString() << std::endl;
        RealNumbers::HCFResult hcf3 = RealNumbers::euclidsAlgorithmHCF(0, 5);
        std::cout << "   HCF(0, 5): " << hcf3.toString() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error in HCF: " << e.what() << std::endl;
    }

    // 3. Prime Factorization
    std::cout << "\n3. Prime Factorization:" << std::endl;
    try {
        RealNumbers::PrimeFactorizationResult pf1 = RealNumbers::getPrimeFactorization(3825);
        std::cout << "   Prime Factors of 3825: " << pf1.toString() << std::endl;
        RealNumbers::PrimeFactorizationResult pf2 = RealNumbers::getPrimeFactorization(140);
        std::cout << "   Prime Factors of 140: " << pf2.toString() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error in Prime Factorization: " << e.what() << std::endl;
    }

    // 4. HCF and LCM Details
    std::cout << "\n4. HCF and LCM Details (using prime factorization for HCF):" << std::endl;
    try {
        RealNumbers::HCFAndLCMResult details1 = RealNumbers::getHCFAndLCMDetails(96, 404);
        std::cout << "   Details for 96 and 404: " << details1.toString() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error in HCF/LCM Details: " << e.what() << std::endl;
    }
     std::cout << "   LCM(96, 404) using direct formula: " << RealNumbers::calculateLCM(96, 404) << std::endl;


    // 5. Irrationality Check for Sqrt(N)
    std::cout << "\n5. Irrationality Check for Sqrt(N):" << std::endl;
    RealNumbers::IrrationalityCheckResult irr1 = RealNumbers::checkSqrtIrrationality(5);
    std::cout << "   " << irr1.toString() << std::endl;
    RealNumbers::IrrationalityCheckResult irr2 = RealNumbers::checkSqrtIrrationality(9);
    std::cout << "   " << irr2.toString() << std::endl;
    RealNumbers::IrrationalityCheckResult irr3 = RealNumbers::checkSqrtIrrationality(6); // Non-prime, non-perfect square
    std::cout << "   " << irr3.toString() << std::endl;

    // 6. Decimal Expansion Type
    std::cout << "\n6. Decimal Expansion Type:" << std::endl;
    try {
        RealNumbers::DecimalExpansionResult dec1 = RealNumbers::getDecimalExpansionType(13, 3125); // Terminating
        std::cout << "   13/3125: " << dec1.toString() << std::endl;
        RealNumbers::DecimalExpansionResult dec2 = RealNumbers::getDecimalExpansionType(1, 7);    // Non-terminating recurring
        std::cout << "   1/7: " << dec2.toString() << std::endl;
        RealNumbers::DecimalExpansionResult dec3 = RealNumbers::getDecimalExpansionType(17, 8);   // Terminating
        std::cout << "   17/8: " << dec3.toString() << std::endl;
        RealNumbers::DecimalExpansionResult dec4 = RealNumbers::getDecimalExpansionType(77, 210); // Simplify to 11/30 (Non-term recurring)
        std::cout << "   77/210: " << dec4.toString() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error in Decimal Expansion: " << e.what() << std::endl;
    }
    
    // 7. Basic Polynomial Analysis (from coefficients)
    std::cout << "\n7. Basic Polynomial Analysis (from coefficients):" << std::endl;
    // x^2 - 3x + 2 => coeffs: [1, -3, 2]
    std::vector<double> coeffs_poly1 = {1.0, -3.0, 2.0};
    RealNumbers::BasicPolynomialInfo info1 = RealNumbers::analyzePolynomialCoeffs(coeffs_poly1);
    std::cout << "   Poly 1 (x^2 - 3x + 2): " << info1.toString() << std::endl;

    // 2x^3 + 0x^2 - 5x + 1 => coeffs: [2, 0, -5, 1]
    std::vector<double> coeffs_poly2 = {2.0, 0.0, -5.0, 1.0};
    RealNumbers::BasicPolynomialInfo info2 = RealNumbers::analyzePolynomialCoeffs(coeffs_poly2);
    std::cout << "   Poly 2 (2x^3 - 5x + 1): " << info2.toString() << std::endl;
    
    // Constant polynomial: 7 => coeffs: [7]
    std::vector<double> coeffs_poly3 = {7.0};
    RealNumbers::BasicPolynomialInfo info3 = RealNumbers::analyzePolynomialCoeffs(coeffs_poly3);
    std::cout << "   Poly 3 (7): " << info3.toString() << std::endl;

    // Leading zero coefficients: 0x^2 + 2x - 1 => effective coeffs {2, -1} for 2x - 1
    std::vector<double> coeffs_poly4 = {0.0, 0.0, 2.0, -1.0};
    RealNumbers::BasicPolynomialInfo info4 = RealNumbers::analyzePolynomialCoeffs(coeffs_poly4);
    std::cout << "   Poly 4 (0x^3 + 0x^2 + 2x - 1): " << info4.toString() << std::endl;
    
    // Zero polynomial
    std::vector<double> coeffs_poly5 = {0.0, 0.0, 0.0};
    RealNumbers::BasicPolynomialInfo info5 = RealNumbers::analyzePolynomialCoeffs(coeffs_poly5);
    std::cout << "   Poly 5 (Zero polynomial): " << info5.toString() << std::endl;


    std::cout << "\n--- End of Real Numbers Examples ---" << std::endl;
    return 0;
}

File: java_code/src/main/java/michu/fr/realnumbers/RealNumbersExamples.java
package michu.fr.realnumbers; // Assuming App.java is here, or place it in michu.fr.realnumbers

// import michu.fr.realnumbers.RealNumberUtils;
import michu.fr.realnumbers.models.*; // Import all models from the realnumbers.models package

import java.util.Map;

public class RealNumbersExamples {

    public static void main(String[] args) {
        System.out.println("--- Real Numbers Examples ---");

        // 1. Euclid's Division Lemma
        System.out.println("\n1. Euclid's Division Lemma:");
        try {
            EuclidLemmaResult lemma1 = RealNumberUtils.euclidsDivisionLemma(455, 42);
            System.out.println("   455 divided by 42: " + lemma1); // Uses toString() of the model
            EuclidLemmaResult lemma2 = RealNumberUtils.euclidsDivisionLemma(10, 3);
            System.out.println("   10 divided by 3: " + lemma2.getEquation()); // Access specific field
        } catch (IllegalArgumentException e) {
            System.err.println("   Error: " + e.getMessage());
        }

        // 2. Euclid's Algorithm for HCF
        System.out.println("\n2. Euclid's Algorithm for HCF:");
        HCFResult hcf1 = RealNumberUtils.euclidsAlgorithmHCF(455, 42);
        System.out.println("   HCF(455, 42): " + hcf1.getHcf());
        HCFResult hcf2 = RealNumberUtils.euclidsAlgorithmHCF(96, 404);
        System.out.println("   HCF(96, 404): " + hcf2.getHcf());
        HCFResult hcf3 = RealNumberUtils.euclidsAlgorithmHCF(17, 0);
        System.out.println("   HCF(17, 0): " + hcf3.getHcf());

        // 3. Prime Factorization
        System.out.println("\n3. Prime Factorization:");
        try {
            PrimeFactorizationResult pf1 = RealNumberUtils.getPrimeFactorization(3825);
            System.out.println("   Prime factors of 3825: " + pf1.getFactors());
            PrimeFactorizationResult pf2 = RealNumberUtils.getPrimeFactorization(96);
            System.out.println("   Prime factors of 96: " + pf2.getFactors());
             // Example showing direct access to factors map
            System.out.print("   Factors of 12: ");
            PrimeFactorizationResult pf12 = RealNumberUtils.getPrimeFactorization(12);
            for (Map.Entry<Integer, Integer> entry : pf12.getFactors().entrySet()) {
                System.out.print(entry.getKey() + "^" + entry.getValue() + " ");
            }
            System.out.println();

        } catch (IllegalArgumentException e) {
            System.err.println("   Error: " + e.getMessage());
        }


        // 4. HCF and LCM (using prime factorization results and direct formula)
        System.out.println("\n4. HCF and LCM Details:");
        try {
            HCFAndLCMResult hcfLcmDetails1 = RealNumberUtils.getHCFAndLCMDetails(96, 404);
            System.out.println("   Details for 96 and 404: " + hcfLcmDetails1);
            
            // Using individual methods
            PrimeFactorizationResult pf96 = RealNumberUtils.getPrimeFactorization(96);
            PrimeFactorizationResult pf404 = RealNumberUtils.getPrimeFactorization(404);
            int hcfFromPf = RealNumberUtils.hcfFromPrimeFactorization(pf96.getFactors(), pf404.getFactors());
            long lcmFromPf = RealNumberUtils.lcmFromPrimeFactorization(pf96.getFactors(), pf404.getFactors());
            System.out.println("   HCF(96,404) from PF: " + hcfFromPf);
            System.out.println("   LCM(96,404) from PF: " + lcmFromPf);

            long lcmDirect = RealNumberUtils.calculateLCM(90, 144);
            System.out.println("   LCM(90, 144) direct: " + lcmDirect);


        } catch (IllegalArgumentException | ArithmeticException e) {
            System.err.println("   Error: " + e.getMessage());
        }

        // 5. Irrationality Check (for sqrt)
        System.out.println("\n5. Irrationality Check (sqrt):");
        IrrationalityCheckResult irr1 = RealNumberUtils.checkSqrtIrrationality(5);
        System.out.println("   sqrt(5): " + irr1);
        IrrationalityCheckResult irr2 = RealNumberUtils.checkSqrtIrrationality(9);
        System.out.println("   sqrt(9): " + irr2);
        IrrationalityCheckResult irr3 = RealNumberUtils.checkSqrtIrrationality(6);
        System.out.println("   sqrt(6): " + irr3);

        // 6. Decimal Expansion Type
        System.out.println("\n6. Decimal Expansion Type:");
        try {
            DecimalExpansionResult de1 = RealNumberUtils.getDecimalExpansionType(13, 3125); // Terminating
            System.out.println("   13/3125: " + de1);
            DecimalExpansionResult de2 = RealNumberUtils.getDecimalExpansionType(64, 455); // Non-terminating recurring
            System.out.println("   64/455: " + de2);
            DecimalExpansionResult de3 = RealNumberUtils.getDecimalExpansionType(17, 8); // Terminating
            System.out.println("   17/8: " + de3);
            DecimalExpansionResult de4 = RealNumberUtils.getDecimalExpansionType(1, 7); // Non-terminating recurring
            System.out.println("   1/7: " + de4);
             DecimalExpansionResult de5 = RealNumberUtils.getDecimalExpansionType(77, 210); // 11/30 -> Non-terminating recurring
            System.out.println("   77/210: " + de5);


        } catch (IllegalArgumentException e) {
            System.err.println("   Error: " + e.getMessage());
        }
        
    }
        
}